src/crypto/kdf/blake2b.rs

```diff
@@ -0,0 +1,163 @@
+//! `blake2b` is the current default key derivation scheme of `libsodium`.
+
+use ffi;
+
+/// Number of bytes in a `Key`.
+pub const KEYBYTES: usize = ffi::crypto_kdf_blake2b_KEYBYTES as usize;
+
+/// Number of bytes in a `Context`
+pub const CONTEXTBYTES: usize = ffi::crypto_kdf_blake2b_CONTEXTBYTES as usize;
+
+/// Minimum number of bytes in a subkey.
+pub const BYTES_MIN: usize = ffi::crypto_kdf_blake2b_BYTES_MIN as usize;
+
+/// Maximum number of bytes in a subkey.
+pub const BYTES_MAX: usize = ffi::crypto_kdf_blake2b_BYTES_MAX as usize;
+
+new_type! {
+    /// `Key` for key derivation.
+    public Key(KEYBYTES);
+}
+
+/// `gen_key()` randomly generates a key for key derivation.
+///
+/// THREAD SAFETY: `gen_key()` is thread-safe provided that you have
+/// called `sodiumoxide::init()` once before using any other function
+/// from sodiumoxide.
+pub fn gen_key() -> Key {
+    use randombytes::randombytes_into;
+
+    let mut key = [0; KEYBYTES];
+    randombytes_into(&mut key);
+    Key(key)
+}
+
+/// `derive_from_key` derives the subkey_id-th subkey from the master key `key` and the context `ctx`
+///
+/// Fails if the length of subkey is not within the bounds given by `BYTES_MIN` and `BYTES_MAX`.
+pub fn derive_from_key(
+    subkey: &mut [u8],
+    subkey_id: u64,
+    ctx: [u8; CONTEXTBYTES],
+    key: &Key,
+) -> Result<(), ()> {
+    unsafe {
+        let r = ffi::crypto_kdf_blake2b_derive_from_key(
+            subkey.as_mut_ptr() as _,
+            subkey.len(),
+            subkey_id,
+            ctx.as_ptr() as *mut i8,
```

Comment:
Can't this just be `ctx.as_ptr()`?

Reply:
The `*mut i8` cast is necessary since `&[u8]::as_ptr` returns `*mut u8` and rust has no notion of type compatibility in this case.

Reply:
Wait a minute. https://doc.rust-lang.org/std/primitive.slice.html#method.as_ptr returnds `*const u8` you cast it to `*mut i8`. Is it legal?

Reply:
There's no good reason for `ctx` to be a cast to a `*mut`, and it's sketchy as hell to cast it to one. The following code typechecks just fine:
```suggestion
            ctx.as_ptr() as *const i8,
```

Reply:
For that matter so does
```suggestion
            ctx.as_ptr() as _,
```

Reply:
> ctx.as_ptr() as *const i8,

Nice

Reply:
I commited your suggestion. I feel slightly ashamed... stupid mistake...